---
title: "(16S amplicon sequencing using PacBio reads) : DADA2R_to_QIIME2"
author: "Nandan Deshpande"
date: "5 Sep 2021"
output:
  html_document:
    df_print: paged
    toc: yes
  html_notebook:
    number_sections: yes
    theme: united
    toc: yes
    toc_float: yes
editor_options:
  chunk_output_type: inline
---
__Last updated__: `r Sys.Date()`

```{r, echo=FALSE}
library(knitr)
opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.width=20, fig.height=8) 
```


# Description
SIH Project: [PIPE-2417](https://ctdshub.atlassian.net/secure/RapidBoard.jspa?rapidView=27&projectKey=PIPE&modal=detail&selectedIssue=PIPE-2417)


# People

__LCI:__ ???

__PhD candidate:__ Dr Giselle Cooke

__USyd Researcher__: Dr Joanna Harnett  


__SIH staff members__

* Tracy Chew (Nandan's group lead)
* Nandan Deshpande 

# Compute facilities

Artemis

Project: Sperm_quality

Directory: /project/RDS-FMH-Sperm_quality-RW/nandanDeshpande/PIPE-2417_Stallion_microbiome/


# PRIMARY REFERENCES
Importing dada2 and Phyloseq objects to QIIME 2: https://forum.qiime2.org/t/importing-dada2-and-phyloseq-objects-to-qiime-2/4683

```{r}

# At command prompt 
module load R/4.0.4 

# Type R and "Enter" at the command prompt to enter R-environment

#Load modules
library(dada2);packageVersion("dada2")
library(Biostrings); packageVersion("Biostrings")
library(ShortRead); packageVersion("ShortRead")
library(ggplot2); packageVersion("ggplot2")
library(reshape2); packageVersion("reshape2")
library(gridExtra); packageVersion("gridExtra")
library(phyloseq); packageVersion("phyloseq")

# SET paths

base_path="/project/RDS-FMH-Sperm_quality-RW/nandanDeshpande/PIPE-2417_Stallion_microbiome/intervention_study"

path2 <- base_path+"/data/raw_data_zipped"
path.out <- base_path+"analysis/approach_2_DADA2R_to_QIIME2_BenjaminCallahan/iteration_2_withChimeraRemoval/Figures"
path.rds <- base_path+"analysis/approach_2_DADA2R_to_QIIME2_BenjaminCallahan/iteration_2_withChimeraRemoval/RDS/"
```

# Read the processed/stored '.rds' file(s)
```{r}

st2 <- readRDS(file.path(path.rds, "Intervention_st2.nochim.rds"))
tax2 <- readRDS(file.path(path.rds, "Intervention_tax2.nochim_Silva128.rds"))
tax2_species_level <- readRDS(file.path(path.rds, "Intervention_tax2_Silva128_withSpecies.rds"))

```

# Prepare and Export Taxonomy, OTU Table
```{r}


# Create a phyloseq object
ps <- phyloseq(otu_table(st2, taxa_are_rows=FALSE),tax_table(tax2))



#Prepare and Export Taxonomy
tax<-as(tax_table(ps),"matrix")
tax_cols <- colnames(tax)
tax<-as.data.frame(tax)
tax$taxonomy<-do.call(paste, c(tax[tax_cols], sep=";"))
for(co in tax_cols) tax[co]<-NULL
write.table(tax, "tax_for_qqime2.txt", quote=FALSE, col.names=FALSE, sep="\t")


# Export feature/OTU table

# As a biom file

#library(biomformat);packageVersion("biomformat")
## [1] ‘1.6.0’

otu<-t(as(otu_table(ps),"matrix")) # 't' to transform if taxa_are_rows=FALSE
#if taxa_are_rows=TRUE
#otu<-as(otu_table(GlobalPatterns),"matrix"))
otu_biom<-make_biom(data=otu)
write_biom(otu_biom,"otu_biom.biom")

# As a text file
write.table(t(st2), "seqtab.txt", sep="\t", row.names=TRUE, col.names=NA, quote=FALSE)
#or from the phyloseq object, 't' to transform if taxa_are_rows=FALSE, no 't' if taxa_are_rows=TRUE
#write.table(t(otu_table(ps), "seqtab.txt",sep="\t", row.names=TRUE, col.names=NA, quote=FALSE)


#uniquesToFasta: Write a uniques vector to a FASTA file : https://www.rdocumentation.org/packages/dada2/versions/1.0.3/topics/uniquesToFasta
uniquesToFasta(st2, fout='st2_rep-seqs.fna', ids=colnames(st2))


# EXIT R environment

```


# In QIIME2
```{r}

# The commands from this point froward should be run outside R environment ; at a command prompt on Artemis/local machine

#If you are running QIIME2 on 

  # Artemis (USyd)
# Load QIIME2 module in artemis
module load anaconda3
conda activate qiime2-2020.8

#else
  #Please install QIIME2 as per https://docs.qiime2.org/2021.4/install/

#Import feature table from exported biom (Check if this worked)
qiime tools import \
  --input-path otu_biom.biom \
  --type 'FeatureTable[Frequency]' \
  --input-format BIOMV100Format \
  --output-path feature-table2_from_DADA2R.qza

#OR

#Import feature table from text file:
echo -n "#OTU Table" | cat - seqtab.txt > seqtab-biom-table.txt


#biom is a command from qiime2 package
biom convert -i seqtab-biom-table.txt -o  seqtab-biom-table.biom --table-type="OTU table" --to-hdf5

qiime tools import \
--input-path  seqtab-biom-table.biom \
--type 'FeatureTable[Frequency]' \
--input-format BIOMV210Format \
--output-path feature-table2_from_DADA2R.qza


## Import the taxonomy table:
qiime tools import \
--type 'FeatureData[Taxonomy]' \
--input-format HeaderlessTSVTaxonomyFormat \
--input-path tax_for_qqime2.txt \
--output-path taxonomy_from_DADA2R.qza

qiime metadata tabulate \
  --m-input-file taxonomy_from_DADA2R.qza \
  --o-visualization taxonomy_from_DADA2R.qzv


## Generate visualization : taxa-bar-plots
qiime taxa barplot \
  --i-table feature-table2_from_DADA2R.qza \
  --i-taxonomy taxonomy_from_DADA2R.qza \
  --m-metadata-file metadata_sheet2.0.tsv \
  --o-visualization taxa-bar-plots_from_DADA2R_metadata_sheet2.0.qzv


## Convert biom format to tab-delimited table format:
biom convert -i table.biom -o table.from_biom.txt --to-tsv

```